### ISSUE TITLE
Bug with avro tool write

### ISSUE THREAD

**Issue by whg517** [mtth/hdfs#133]
Use of HDFSCLI with the [avro tool](http://avro.apache.org/docs/current/gettingstartedpython.html) with the following problems:

code:

```python
import avro
from avro.datafile import DataFileWriter, DataFileReader
from avro.io import DatumWriter

from hdfs.ext.kerberos import KerberosClient

client = KerberosClient('http://m1.node.hadoop:50070')

schema = {"namespace": "example.avro",
          "type": "record",
          "name": "User",
          "fields": [
              {"name": "name", "type": "string"},
              {"name": "favorite_number", "type": ["int", "null"]},
              {"name": "favorite_color", "type": ["string", "null"]}
          ]}

schema = avro.schema.SchemaFromJSONData(schema)

records = [
    {"name": "Alyssa", "favorite_number": 256},
    {"name": "Ben", "favorite_number": 7, "favorite_color": "red"}
]


def avro_writer(path):
    with client.write(path, overwrite=True) as writer:
        with DataFileWriter(writer, DatumWriter(), schema) as avro_writer:
            for record in records:
                avro_writer.append(record)


def avro_reader(path):
    with client.read(path) as reader:
        with DataFileReader(reader, DatumWriter()) as avro_reader:
            for i in avro_reader:
                print(i)


if __name__ == '__main__':
    """"""
    avro_writer('/tmp/test.avro')

```

1. write

It can be written successfully, but an error is reported

```
Traceback (most recent call last):
  File "/home/kevin/workspaces/develop/python/StudyPython/third_library/avro_/d3.py", line 42, in <module>
    avro_writer('/tmp/test.avro')
  File "/home/kevin/workspaces/develop/python/StudyPython/third_library/avro_/d3.py", line 30, in avro_writer
    avro_writer.append(record)
  File "/home/kevin/.virtualenvs/StudyPython-AS8qU0JI/lib/python3.6/site-packages/avro/datafile.py", line 205, in __exit__
    self.close()
  File "/home/kevin/.virtualenvs/StudyPython-AS8qU0JI/lib/python3.6/site-packages/avro/datafile.py", line 327, in close
    self.writer.close()
AttributeError: 'AsyncWriter' object has no attribute 'close'

```

2. read

Unable to read properly

```
Traceback (most recent call last):
  File "/home/kevin/workspaces/develop/python/StudyPython/third_library/avro_/d3.py", line 42, in <module>
    avro_reader('/tmp/test.avro')
  File "/home/kevin/workspaces/develop/python/StudyPython/third_library/avro_/d3.py", line 35, in avro_reader
    with DataFileReader(reader, DatumWriter()) as avro_reader:
  File "/home/kevin/.virtualenvs/StudyPython-AS8qU0JI/lib/python3.6/site-packages/avro/datafile.py", line 351, in __init__
    self._read_header()
  File "/home/kevin/.virtualenvs/StudyPython-AS8qU0JI/lib/python3.6/site-packages/avro/datafile.py", line 459, in _read_header
    self.reader.seek(0, 0)
io.UnsupportedOperation: seek

```

Can you fix it ? 
Thank you.

**Comment by mtth** [comment]
Can you try with the [Avro extension](https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.avro)? It supports reading and writing Avro files directly from HDFS, with a slightly more concise API.

**Comment by whg517** [comment]
> Can you try with the [Avro extension](https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.avro)? It supports reading and writing Avro files directly from HDFS, with a slightly more concise API.

Hello,
Excuse me again.

[Avro extension](https://hdfscli.readthedocs.io/en/latest/api.html#module-hdfs.ext.avro) cannot meet the expected requirements.
As you said in [Avro format read-write support for external schemas](https://github.com/mtth/hdfs/issues/132#issuecomment-506977090) , using external schemas is beyond the scope of the library. So I chose [avro python](http://avro.apache.org/docs/current/gettingstartedpython.html) , but when both were written to HDFS, it reported an error and failed to read.

Do you have a plan to support [avro python](http://avro.apache.org/docs/current/gettingstartedpython.html)  ?

Thank you for your reply.

**Comment by mtth** [comment]
> Avro extension cannot meet the expected requirements.

What is missing?

> As you said in Avro format read-write support for external schemas , using external schemas is beyond the scope of the library. 

To clarify: integrating with an external schema provider (e.g. a schema registry) is out of scope, using explicit user-defined schemas isn't. For instance [`AvroReader`](https://hdfscli.readthedocs.io/en/latest/api.html#hdfs.ext.avro.AvroReader) already accepts an optional schema as input.

**Comment by whg517** [comment]
Sorry, I described the error, but what I really meant was to pass in a user-defined schema, rather than integrate the schema provider into this plug-in. And I re-read the AvroWriter documentation, which does support the `schema` parameter.


Can you make `AvroReader` to also support `schema` ?

I see that `fastavro` is used by your Avro Extensions and [fastavro.read](https://fastavro.readthedocs.io/en/latest/reader.html#fastavro._read_py.reader) supports user-defined schemas.

**Scene description:**

When my schema was at version v1, I wrote the following data

```json
{
    "namespace": "example.avro",
    "type": "record",
    "name": "User",
    "fields": [
        {"name": "name", "type": "string"},
        {"name": "favorite_number", "type": ["int", "null"]}
    ]
}
```

When my schema evolves to v2 due to requirements, add some fields and add a 'default' default value according to the avro rule.

```json
{
    "namespace": "example.avro",
    "type": "record",
    "name": "User",
    "fields": [
        {"name": "name", "type": "string"},
        {"name": "favorite_number", "type": ["int", "null"]},
        {"name": "favorite_color", "type": "string", "default": "red"}
    ]
}
```

-------

When I read the data written by the v1 version, the field and default values are automatically added.

Of course, this problem has been solved, because I can also call fastavro to implement the corresponding function.

Excuse me.

**Comment by mtth** [comment]
> Can you make AvroReader to also support schema ?

I'll try to get that in shortly.

**Comment by mtth** [comment]
FYI, just released `2.5.8` which adds a `reader_schema` option to [`AvroReader`](https://hdfscli.readthedocs.io/en/latest/api.html#hdfs.ext.avro.AvroReader).
